contracts/token/ERC721/ConfidentialERC721URIStorage.sol

```diff
@@ -35,31 +39,79 @@ abstract contract ConfidentialERC721URIStorage is IERC165, ConfidentialERC721 {
     function _setTokenURI(
         address to,
         uint256 tokenId,
-        gtUint64 _tokenURI
+        ctUint64[] calldata _itTokenURI,
+        bytes[] calldata _itSignature
     ) internal virtual {
-        _tokenURIs[tokenId] = MpcCore.offBoardCombined(_tokenURI, to);
+        gtUint64[] memory _tokenURI = new gtUint64[](_itTokenURI.length);
+
+        itUint64 memory it;
+
+        for (uint256 i = 0; i < _itTokenURI.length; ++i) {
+            it.ciphertext = _itTokenURI[i];
+            it.signature = _itSignature[i];
+
+            _tokenURI[i] = MpcCore.validateCiphertext(it);
+        }
+
+        _setTokenURI(to, tokenId, _tokenURI);
+    }
+
+    /**
+     * @dev Sets `_tokenURI` as the tokenURI of `tokenId`.
+     *
+     * Emits {MetadataUpdate}.
+     */
+    function _setTokenURI(
+        address to,
+        uint256 tokenId,
+        gtUint64[] memory _tokenURI
+    ) private {
+        if (!isMinted(tokenId)) {
+            revert ERC721URIStorageNonMintedToken(tokenId);
+        }
+
+        ctUint64[] memory userTokenURI = new ctUint64[](_tokenURI.length);
+        ctUint64[] memory networkTokenURI = new ctUint64[](_tokenURI.length);
+
+        utUint64 memory offBoardCombined;
+
+        for (uint256 i = 0; i < _tokenURI.length; ++i) {
+            offBoardCombined = MpcCore.offBoardCombined(_tokenURI[i], to);
+
+            userTokenURI[i] = offBoardCombined.userCiphertext;
```

Comment:
why not to save it also in ut form instead of userToken and networkToken

Reply:
Solidity doesn't support mapping to array of structs (one struct for each character in the array).

I thought about creating a new struct for UT String and adding it to the MPC Core as well. If Solidity supports mapping to struct that holds an array, this would be a nice design.